app/src/main/java/com/eveningoutpost/dexdrip/cgm/carelinkfollow/client/CareLinkClient.java

```diff
@@ -0,0 +1,656 @@
+package com.eveningoutpost.dexdrip.cgm.carelinkfollow.client;
+
+import android.util.Log;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import okhttp3.*;
+
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.ActiveNotification;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.ClearedNotification;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.ConnectData;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.ConnectDataResult;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.CountrySettings;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.Marker;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.MonitorData;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.Profile;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.RecentData;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.SensorGlucose;
+import com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.User;
+import com.google.gson.*;
+
+
+/**
+ * Medtronic CareLink Follower Client
+ *   - API client for communication with CareLink
+ */
+public class CareLinkClient {
+    protected String carelinkUsername;
+    protected String carelinkPassword;
+    protected String carelinkCountry;
+    protected static final String CARELINK_CONNECT_SERVER_EU = "carelink.minimed.eu";
+    protected static final String CARELINK_CONNECT_SERVER_US = "carelink.minimed.com";
+    protected static final String CARELINK_LANGUAGE_EN = "en";
+    protected static final String CARELINK_LOCALE_EN = "en";
+    protected static final String CARELINK_AUTH_TOKEN_COOKIE_NAME = "auth_tmp_token";
+    protected static final String CARELINK_TOKEN_VALIDTO_COOKIE_NAME = "c_token_valid_to";
+    protected static final int AUTH_EXPIRE_DEADLINE_MINUTES = 1;
+
+    protected static final SimpleDateFormat[] ZONED_DATE_FORMATS = {
```

Comment:
Did you think about my previous comment https://github.com/NightscoutFoundation/xDrip/pull/1649#discussion_r958811484?

Reply:
During the last 1.5 years I have found a lot of different exotic date/time formats in the CareLink JSON messages depending on the device (Guardian Connect CGM / 7xxG pump), platform (IOS application / Android application) and region (US and outside US). I have added all of them to the zoned formats and tested them. I am not sure if the proposed class will be suitable and able to handle all cases. I can test maybe some scenarios, but not all of them.

Reply:
Then it would be best to add all those cases to https://github.com/tolot27/xDrip/blob/master/app/src/test/java/com/eveningoutpost/dexdrip/Models/DateUtilTest.java#L27 and we can be sure that the unit tests will find them.

It is highly recommend to create unit tests. I can test your listed cases against our existing class, if you want.

Reply:
I have checked the `tolerantFromISODateString` method and although it works on all the date/time with time zone information that I have previously found, I still have issues with it:
1. It returns a date even if there is no time zone information. For example: 2022-01-01T20:22:01 -> 2022-01-01T21:22:01.000+0100 This is not good for me, because it is a different scenario if no time zone information is available. In this case I try to extract the time zone from another date field of the CareLink data and add this information to the other date fields, which doesn't have time zone (this is a rare scenario, but it can happen in special cases). On the other hand it fails if the there is no time zone, but there are milliseconds: 2022-01-01T20:22:01.012 -> Exception
2. As I wrote earlier I have written the CareLink Client as an independent java library, it has a separate repo. I wouldn't like to add dependencies on xdrip classes in the core CareLink Client.

Reply:
For 1.: I've fixed the exception you mentioned with #2412.
I believe you are wrong regarding to the added time zone. The function `tolerantFromISODateString` parses an ISO 8601 formatted string w/o time zone information and returns a Java `Date` object. `Date` contains just milliseconds since 1970-01-01 and is UTC. I think the time zone you get is if you convert it back to a date/time string. In this case, your local timezone is added. If you have no time zone information you can get it from somewhere else or have to stick at UTC.

Reply:
I can't remember in which CareLink uploader app or server scenario (7xxG pump or GuardianConnect / region / IOS or Android), but some datetime strings doesn't have timezone. In this case the timezone can be extracted from another datetime field, which always have timezone information and this timezone should be used for all datetime strings without timezone information, which is probably the uploader app's timezone. That was the reason for creating the `ZONED_DATE_FORMATS`, which only parses datetime strings with timezone. So even if I use the `tolerantFromISODateString`, I would still have to check first for the existence of the timezone in the datetime string using the `ZONED_DATE_FORMATS`, so at the end I would be using two different methods for the same purpose: parse datetime with timezone.
So you would like me to use the `tolerantFromISODateString` for parsing the datetime string and the  `ZONED_DATE_FORMATS` only to check if timezone is provided?

Reply:
I just wanted to emphasize that `tolerantFromISODateString` does not add a time zone. You should be careful if you convert a `Date` object to an ISO formatted date string. The `DateUtil` class provides a method for it for NS uploads.

You mentioned that other date/time fields always contain timezone information. Why don't you use these information?

Anyway, we can resolve this review comment, because it is know just academic and does not have any influence on this PR.